### PR TITLE
ceph: add rook-version and ceph-version tags to crash deployment

### LINKED ISSUE #4357: Crashcollector deployments should have Rook & Ceph version labels
The new crashcollector deployments should get `rook-version` and `ceph-version` labels. This should be a pretty quick, easy fix that we should get in before v1.2 is released.

### PR DESCRIPTION
**Pull request by egafford**
**Description of your changes:**
This change adds rook-version and ceph-version tags to crash
collector deployments, in keeping with other Ceph deployments.

**Which issue is resolved by this Pull Request:**
Resolves #4357
Signed-off-by: Elise Gafford <egafford@redhat.com>

**Checklist:**

- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [ ] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](https://github.com/rook/rook/blob/master/CONTRIBUTING.md#comments)
- [ ] Add the flag for skipping the CI if this PR does not require a build. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for more details.

[test ceph]

### CODE REVIEW

**Comment by egafford** [comment]
Note: Logic around which ceph-version label to use if multiple versions are detected in cluster deployments may be incorrect; I'd appreciate eyes on that from folks with more detailed knowledge of the upgrade process. Thanks.

File: pkg/operator/ceph/cluster/crash/reconcile.go [hunk 1]
```diff
@@ -159,3 +167,42 @@ func (r *ReconcileNode) cephPodList() ([]corev1.Pod, error) {
 
 	return cephPods, nil
 }
+
+func (r *ReconcileNode) getCephVersion(cephCluster cephv1.CephCluster) (*version.CephVersion, error) {
```
Comment (BlaineEXE) [inline pkg/operator/ceph/cluster/crash/reconcile.go]: This should be unnecessary. The Ceph version on the label should be the Ceph version of the image being used, the one detected at operator start time.

Comment (BlaineEXE) [reply pkg/operator/ceph/cluster/crash/reconcile.go]: This might be less straightforward than I was hoping given how `register_controllers.go` is written, but in theory this should look something like starting the object or file controller. `FilesystemController` for example has the Ceph version from `clusterInfo`. @leseb wrote this code, so he might have a better idea of how to properly patch that in.

Comment (egafford) [reply pkg/operator/ceph/cluster/crash/reconcile.go]: Right; I felt like there should be a saner way to do this, but since the reconcile loop iterates over many clusters (conceivably at different versions) we do need to scrape it from somewhere. If there is an object I can spin up, that'd be great, but given that the deployment is created inside the reconcile loop this was the least baroque way I saw to implement with what I had in scope. (The other I saw was to find a pod with an image matching the crash-collector's, then map that back to its deployment, but that seemed more cumbersome.)

If there is a cleaner way to implement, I'd be quite glad of it; I agree that this is a lot (and I do have concerns about the viability of this approach during an upgrade process).

Comment (BlaineEXE) [reply pkg/operator/ceph/cluster/crash/reconcile.go]: The ceph version should always be the version of Ceph which is running in the pod's container. I think the most reliable and correct thing to do would be to run the job to detect the ceph version of the image before updating the crashcontroller.

Comment (egafford) [reply pkg/operator/ceph/cluster/crash/reconcile.go]: Okay; I thought of doing that, but I saw two problems.

First, since the deployment is created in the reconcile loop, it may not yet have been created, In this case we'd have to either spawn a new detect-version pod just for this deployment or have special logic to leave the version blank until a crash-collector pod had spawned under the deployment.

Second, it seemed a bit heavy from a resource standpoint. Still, I was considering it as though spawning a new detect-version pod would be needed (potentially for every deployment on every reconcile pass). If we can run the job without spawning the pod, that's a lot saner.

Thoughts?

Comment (leseb) [reply pkg/operator/ceph/cluster/crash/reconcile.go]: Sorry for the late reply.

> Right; I felt like there should be a saner way to do this, but since the reconcile loop iterates over many clusters (conceivably at different versions) we do need to scrape it from somewhere.

Where do you see that? The crash-collector is a namespaced resource and Rook only supports a single cluster per namespace. We even have this check here: https://github.com/rook/rook/blob/master/pkg/operator/ceph/cluster/crash/reconcile.go#L115-L118 if something is weird.

> If there is an object I can spin up with a proper version.CephVersion, that'd be great, but given that the deployment is created inside the reconcile loop this was the least baroque way I saw to implement with what I had in scope. (The other I saw was to find a pod with an image matching the crash-collector's, then map that back to its deployment, but that seemed more cumbersome.)
> 
> If there is a cleaner way to implement, I'd be quite glad of it; I agree that this is a lot (and I do have concerns about the viability of this approach during an upgrade process).

Can we try to register the ceph version based on the label when we find Ceph pod around here: https://github.com/rook/rook/blob/master/pkg/operator/ceph/cluster/crash/reconcile.go#L122

So we can do something like:

```golang
		for _, cephPod := range cephPods {
			if cephPod.Spec.NodeName == request.Name {
				logger.Debugf("cephPod.Spec.NodeName is %q and request.Name is %q", cephPod.Spec.NodeName, request.Name)
				hasCephPods = true
				for _, podToleration := range cephPod.Spec.Tolerations {
					// Add toleration to the map
					uniqueTolerations.Add(podToleration)
				}
				cv := cephPod.Labels["ceph-version"]
			}
		}
```

And then you can pass `cv` to `createOrUpdateCephCrash()`.

How does that sound?

Comment (leseb) [reply pkg/operator/ceph/cluster/crash/reconcile.go]: Or maybe better, you can add a new filed to the reconcile struct for the ceph version. Wether you prefer.

Comment (travisn) [reply pkg/operator/ceph/cluster/crash/reconcile.go]: Overall I see the design being that we should have a way for controllers to store state that another controller can consume. In this case, the CephCluster controller is going to detect the version of the image and store it somewhere in memory such that the crash controller can read that version. The main thing to be careful of is that this is done safely across goroutines. 

@leseb What is the upgrade path for the crash pods? Does the crash controller also watch for changes to the CephCluster CR so it knows when to update them with the new image? 

How about the following sequence:
-  The CephCluster is created with Ceph version `14.2.4`. Mons, OSDs, and other daemons are created.
- The CephCluster controller saves `cephVersions[image] = 14.2.4` in a package variable
-  The crash controller sees that daemons were created on some nodes and starts the crash pods on those nodes. The version label is added to the deployments since the controller can look up the value in `cephVersions[image]`. 

Next, let's say the CephCluster CR is updated to `14.2.5`:
- The CephCluster controller kicks off an orchestration. The new version is detected and ceph daemons are restarted.
- The crash controller notices the updated CephCluster image. The controller waits until `cephVersions[image]` is populated with the detected version for the new image, then proceeds to update the crash deployments with the new labels and image.

**Comment by egafford** [comment]
@leseb @travisn 

On the last patch, I'm thinking through whether we need to throw up a mutex around any detectCephVersion call (and its subsequent map storage). The case we'd be guarding against would be:

-  goroutine 1 detects version 1
- the image of precisely the same name is updated to Ceph version 2
- goroutine 2 detects version 2
- goroutine 2 registers version 2
- goroutine 1 wakes up from its 100-year slumber to register version 1

This seems reasonably unlikely, and as detectCephVersion is a long-running, cluster-scoped process, I'd much prefer to avoid if at all possible. It also only occurs in cases in which an image name is reused with a different ceph version. Still, it is a conceivable error case, and it would take another update to any cluster using that image before it would be repaired (it wouldn't just be picked up on the next reconcile loop). Locking only around the map put is insufficient to address it. 

Thinking further through, even with any lock we would run the risk of the following scenario:
- version of a crash-collector is running image version 1 and labelled as such
- the image of precisely the same name is updated to Ceph version 2
- on the next reconcile loop, the crash-collector label is updated to Ceph version 2 without necessarily being upgraded
The best way I can see to deal with this would be to make the map a map[clusterName]map[image]version, so that each cluster would need to be upgraded to the new version for the change to propagate to its crash-collector. That's reasonably cumbersome, though.

To (potentially) cut through all this, is it currently documented that the user must provide a new image name for any new image? If not, should it be?

**Comment by BlaineEXE** [comment]
How does the crash collector controller currently know which Ceph image to use when creating its pods? That information has to be passed to it somehow. When that information is passed to it, can we also include the detected Ceph version with it?

**Comment by egafford** [comment]
It's passed as a field of the cephCluster: the struct is stored as cephCluster.Spec.CephVersion. which has two fields: {Image: string, AllowUnsupported bool}. We could change the spec of the CephCluster resource. Still, that's a non-trivial API change, and would need to be a get-only field (if we allow the user to specify version label and image, we're pretty uselessly inviting mismatch). Still, it's a real option if we're willing to accept an API change for this case.

**Comment by travisn** [comment]
The only producer of the version info will be the cluster controller. Nobody else would be starting the detect version job. There will only be one goroutine at a time executing the job, so we don't need to lock it for the purposes of updating the version map.

I don't see the need to support an image of the same name/tag with a different version of Ceph. Our upgrade path requires a new image name if you want your pods to be restarted when an update is made. So I'd suggest we keep it simple as far as mapping a given tag to a given version.

We wouldn't want to add the detected version to the cluster CR next to the image name. That's reserved for user's desired state rather than info that is detected at runtime. But it does sound like a good idea to pass the version as a parameter to the crash controller next to the cluster CR. Then the crash controller can go ahead and update the pods along with the new version that was detected. If the cluster is passing this info down to the crash controller with the cluster CR, we probably don't need a lock around that info since the cluster controller would define when to notify the crash controller. But I'm not sure how the crash controller gets the cluster CR from the cluster controller, so the approach would depend on that since it's certainly a separate goroutine.

**Comment by egafford** [comment]
TL;DR: If one image name always maps to one version, then we never had any concurrency problems. This is great! We do have storing global state in working memory problems, though.

I put one more patch up: I realized that since the map is being stored in process memory, a failed version lookup should be non-fatal error for the crash-collector reconciliation. (This way if the operator bounces, reconciliation of the crash-collector isn't blocked.) If we're willing to live with this for 1.2 with an issue to address it and plans to backport, we're good. Otherwise I'll try to address it tonight.

@travisn, @leseb can validate this, but passing the cephVersion to the crash controller is non-trivial. The crash-controller manager is only added via AddToManagerFuncs; its reconcile loop and that of the Cluster CR are pretty entirely disjoint. The crash-controller manager actually queries k8s for CephCluster CRs as part of its reconcile loop, rather than ever having them passed on the callstack. (Otherwise this would have been done a bit ago. I remember the good old days when @BlaineEXE marked this as a trivial issue).

**Comment by travisn** [comment]
> I put one more patch up: I realized that since the map is being stored in process memory, a failed version lookup should be a non-fatal error for the crash-collector reconciliation. (This way if the operator bounces, reconciliation of the crash-collector isn't blocked.) If we're willing to live with this for 1.2 with an issue to address it and plans to backport, we're good. Otherwise I'll try to address it tonight.

Sounds good to make this a non-fatal error. The crash collector could still be created without these tags.

> @travisn, @leseb can validate this, but passing the cephVersion to the crash controller is non-trivial. The crash-controller manager is only added via AddToManagerFuncs; its reconcile loop and that of the Cluster CR are pretty entirely disjoint. The crash-controller manager actually queries k8s for CephCluster CRs as part of its reconcile loop, rather than ever having them passed on the callstack. (Otherwise this would have been done a bit ago. I remember the good old days when @BlaineEXE marked this as a trivial issue).

Oh right now I remember as well that crash controller is just looking up the cluster CR directly. In this case we would need the lock in order to check for the version that was detected from the cluster controller goroutine. When the crash controller checks for the image in the cephcluster, it would look up in memory what the version is for that image.

**Comment by egafford** [comment]
@travisn, I'm a little confused on why we need a lock on the image: version map. If we're positing that any given image name will always map to one CephVersion, then the global var map put will never actually effect change after the first pass on any one image. We only start to create the  crash-collector after the cephCluster CR exists, and this can only occur after the detect-ceph-version job has completed. Even if there were a race condition we've somehow missed, we'd just label the crash-collector pod with the ceph version on its next reconcile pass.

Am I fundamentally missing something about concurrency in Go package variable usage here (quite possible coming from Python), or are you seeing an error condition I'm not? I can definitely see error conditions if the ceph version for an image changes, as above, but we're positing that this is a user error case (and could hopefully be remedied by updating with the same image under a different name).

**Comment by travisn** [comment]
> @travisn, I'm a little confused on why we need a lock on the image: version map. If we're positing that any given image name will always map to one CephVersion, then the global var map put will never actually effect change after the first pass on any one image. We only start to create the crash-collector after the cephCluster CR exists, and this can only occur after the detect-ceph-version job has completed. Even if there were a race condition we've somehow missed, we'd just label the crash-collector pod with the ceph version on its next reconcile pass.

@egafford There are two scenarios I'm thinking of:
1. After a cluster is configured the operator is restarted. Both the CephCluster and crash collector reconcile loops will start in parallel. Since it takes at least several seconds to detect the ceph version from running a job, I'm assuming the crash collector will commonly finish its initial reconciliation first. It wouldn't find the version in the map, and we would also need a lock for the map due to this happening in parallel.
1. When the CephCluster CR is updated, the crash collector should update its pods with the new image. I'm assuming this would also happen in parallel of the cluster controller detecting the new version.

In both of these cases I'm thinking the crash controller should wait and keep polling the map until the expected version has been discovered. After some reasonable timeout (perhaps 5 minutes) it could continue without the version if not detected yet.

File: pkg/operator/ceph/cluster/crash/reconcile.go [hunk 2]
```diff
@@ -117,6 +118,12 @@ func (r *ReconcileNode) reconcile(request reconcile.Request) (reconcile.Result,
 			logger.Errorf("more than one CephCluster found in the namespace %q, choosing the first one %q", namespace, cephCluster.GetName())
 		}
 
+		clusterImage := cephCluster.Spec.CephVersion.Image
+		cephVersion, err := version.GetImageVersion(clusterImage)
```
Comment (leseb) [inline pkg/operator/ceph/cluster/crash/reconcile.go]: The latest approach looks good and just like @travisn said, a bit of polling wouldn't hurt in the case of the operator restarting.

File: pkg/operator/ceph/cluster/crash/reconcile.go [hunk 3]
```diff
@@ -117,6 +118,12 @@ func (r *ReconcileNode) reconcile(request reconcile.Request) (reconcile.Result,
...
+		if err != nil {
+			logger.Warningf("could not identify ceph version for image %q used by cluster %q: %v", clusterImage, cephCluster.Name, err)
```
Comment (leseb) [inline pkg/operator/ceph/cluster/crash/reconcile.go]: Shouldn't we return here instead of passing an empty version?

File: pkg/operator/ceph/cluster/crash/reconcile.go (hunk 2 above)
Comment (egafford) [reply pkg/operator/ceph/cluster/crash/reconcile.go]: Agreed; I think the word "lock" above made me think "mutex" rather than "retry loop". Implementing a poll.

File: pkg/operator/ceph/cluster/crash/reconcile.go (hunk 3 above)
Comment (egafford) [reply pkg/operator/ceph/cluster/crash/reconcile.go]: There was some agreement above that the lack of version tag should be a non-fatal error. I'm open to changing this if there's concern, though.

File: pkg/operator/ceph/cluster/crash/reconcile.go (hunk 2 above)
Comment (travisn) [reply pkg/operator/ceph/cluster/crash/reconcile.go]: The lock I'm thinking of would involve a simple mutex that is only used when updating the map or reading from the map, thus preventing any invalid access while the map is being updated.

File: pkg/operator/ceph/cluster/crash/reconcile.go (hunk 3 above)
Comment (leseb) [reply pkg/operator/ceph/cluster/crash/reconcile.go]: Sorry I missed this https://github.com/rook/rook/pull/4366/files#diff-340a35fd504236d0f1e7840bac0c6460R90 in my review. All good.

File: pkg/operator/ceph/cluster/crash/reconcile.go [hunk 4]
```diff
@@ -117,6 +125,12 @@ func (r *ReconcileNode) reconcile(request reconcile.Request) (reconcile.Result,
 			logger.Errorf("more than one CephCluster found in the namespace %q, choosing the first one %q", namespace, cephCluster.GetName())
 		}
 
+		cephVersion, err := getImageVersion(cephCluster.Spec.CephVersion.Image)
+		if err != nil && !getVersionFailed {
+			logger.Warningf("ceph version note found for cluster %q: %v", cephCluster.Name, err)
```
Comment (travisn) [inline pkg/operator/ceph/cluster/crash/reconcile.go]: typo: not

File: pkg/operator/ceph/cluster/crash/reconcile.go [hunk 5]
```diff
@@ -159,3 +173,20 @@ func (r *ReconcileNode) cephPodList() ([]corev1.Pod, error) {
 
 	return cephPods, nil
 }
+
+func getImageVersion(image string) (*version.CephVersion, error) {
+	retries := 0
+	for {
```
Comment (travisn) [inline pkg/operator/ceph/cluster/crash/reconcile.go]: Even though it will be functionally equivalent, i'd avoid the appearance of an infinite loop. For example, in simplified terms:
```
for i:=0; i < maxRetries; i++ {
  if successful {
    return version, nil
  } 
}
getVersionFailed = true
return nil, fmt.Errorf("ceph version not found for image %q", image)
```

File: pkg/operator/ceph/cluster/crash/reconcile.go [hunk 6]
```diff
@@ -41,8 +43,14 @@ import (
...
 
+const (
+	getVersionRetryInterval = 5
+	getVersionMaxRetries    = 60
+)
+
 var (
-	logger = capnslog.NewPackageLogger("github.com/rook/rook", controllerName)
+	logger           = capnslog.NewPackageLogger("github.com/rook/rook", controllerName)
+	getVersionFailed = true
```
Comment (travisn) [inline pkg/operator/ceph/cluster/crash/reconcile.go]: Why keep track of getVersionFailed? Is that for testing?

File: pkg/operator/ceph/version/version.go [hunk 7]
```diff
@@ -293,3 +295,15 @@ func ValidateCephVersionsBetweenLocalAndExternalClusters(localVersion, externalV
 
 	return nil
 }
+
+func RegisterImageVersion(image string, version CephVersion) {
+	imageToVersionMap[image] = version
+}
+
+func GetImageVersion(image string) (*CephVersion, error) {
+	version, ok := imageToVersionMap[image]
```
Comment (travisn) [inline pkg/operator/ceph/version/version.go]: Since `RegisterImageVersion()` is called from a separate goroutine from `GetImageVersion()`, we do need a mutex protecting the `imageToVersionMap`

File: pkg/operator/ceph/cluster/crash/reconcile.go (hunk 2 above)
Comment (egafford) [reply pkg/operator/ceph/cluster/crash/reconcile.go]: @travisn Ah, got it; Go 1.6 requires a lock for concurrent read and write. That explains the feedback, and why the book I studied didn't suggest the pattern in this patch would be a problem. Thank you. Updating.

(https://stackoverflow.com/questions/36167200/how-safe-are-golang-maps-for-concurrent-read-write-operations)

Comment (egafford) [reply pkg/operator/ceph/cluster/crash/reconcile.go]: Ahh, got it. Go does require a lock for concurrent read and write on map access.  Updating.

File: pkg/operator/ceph/cluster/crash/reconcile.go (hunk 4 above)
Comment (egafford) [reply pkg/operator/ceph/cluster/crash/reconcile.go]: I'll change the name of this variable: it actually reflects whether or not the get version check has already gone through a full recheck cycle and failed. This bool is intended to prevent warnings showing up on every single reconcile loop until the operator restarts if the version check fails. The loop will continue to look for a version, and if it finds one it will use it, but we'll only write to the logs once about it rather than spamming.

File: pkg/operator/ceph/cluster/crash/reconcile.go (hunk 5 above)
Comment (egafford) [reply pkg/operator/ceph/cluster/crash/reconcile.go]: Fair enough. I was following a pattern elsewhere in the codebase, but I'm happy to do this instead.

File: pkg/operator/ceph/cluster/crash/reconcile.go (hunk 6 above)
Comment (egafford) [reply pkg/operator/ceph/cluster/crash/reconcile.go]: It's to avoid spamming the logs on every reconcile loop if the detect-version job fails. I can remove this if we'd *prefer* to spam the logs on every reconcile loop.

Comment (travisn) [reply pkg/operator/ceph/cluster/crash/reconcile.go]: If it's just for the warning message, I'd think it's ok to print on every reconciliation. It should be an intermittent issue anyway and not be a verbose error.

File: pkg/operator/ceph/cluster/crash/reconcile.go (hunk 4 above)
Comment (travisn) [reply pkg/operator/ceph/cluster/crash/reconcile.go]: Actually I was just referring to "note" --> "not". :)

File: pkg/operator/ceph/cluster/crash/reconcile.go (hunk 6 above)
Comment (egafford) [reply pkg/operator/ceph/cluster/crash/reconcile.go]: Alright; I'll remove this. Simplifies the code, certainly, and if we're okay with the operator seeing this for a bit that's fine. It will mean that in the case that the detect-version job fails consistently for any reason, crash-collector reconciliations will always take a full 5 minutes, but at that point the cluster likely has larger problems.

File: pkg/operator/ceph/version/version.go [hunk 8]
```diff
@@ -293,3 +297,18 @@ func ValidateCephVersionsBetweenLocalAndExternalClusters(localVersion, externalV
 
 	return nil
 }
+
+// RegisterImageVersion stores the CephVersion detected for a specified image for global access.
+func RegisterImageVersion(image string, version CephVersion) {
+	imageToVersionMapLock.Lock()
```
Comment (travisn) [inline pkg/operator/ceph/version/version.go]: Generally speaking, golang recommends using `defer` to ensure a lock is released in case there is a panic in the middle of it. For example:
```
imageToVersionMapLock.Lock()
defer imageToVersionMapLock.Unlock()
imageToVersionMap[image] = version
```

File: pkg/operator/ceph/cluster/crash/reconcile.go [hunk 9]
```diff
@@ -117,6 +124,12 @@ func (r *ReconcileNode) reconcile(request reconcile.Request) (reconcile.Result,
 			logger.Errorf("more than one CephCluster found in the namespace %q, choosing the first one %q", namespace, cephCluster.GetName())
 		}
 
+		clusterImage := cephCluster.Spec.CephVersion.Image
+		cephVersion, ok := getImageVersion(clusterImage)
+		if !ok {
+			logger.Warningf("ceph version not found for image %q used by cluster %q: %v", clusterImage, cephCluster.Name, err)
```
Comment (BlaineEXE) [inline pkg/operator/ceph/cluster/crash/reconcile.go]: Nit:
```suggestion
			logger.Warningf("ceph version not found for image %q used by cluster %q. %v", clusterImage, cephCluster.Name, err)
```

**Comment by travisn** [comment]
> I appreciate that the code is much simpler now compared to previous iterations, but I am still opposed to using a globally scoped map for determining the information in the long term. This sets a bad precedent for others to follow.

@BlaineEXE Do you have another suggestion for the inter-controller communication? Another approach that comes to mind is that the CephCluster controller could store the detected version in the `status` section of the CephCluster CR. Then the other controllers could look there for the detected version while they're already looking for the image. It's a bit more work, but also feels cleaner. But I'd suggest we save that for a future PR, perhaps as we start moving the other controllers (CephObjectStore, CephFilesystem, CephBlockPool, etc) to the controller runtime as well.

**Comment by egafford** [comment]
I admit that I share @BlaineEXE's concerns. I was surprised when the package var map solution was suggested; as I understand the Go idiom it is to be avoided (unlike in Python). This instance is reasonably safe, as there is one writer and KVPs won't change except in case of user error, but if it were taken as precedent to store a truly dynamic map or slice that would be dangerous.

I've basically been waiting until this patch lands and this conversation closes to file an issue for a larger change to replace my own work. If we do have room to store arbitrary data in `status` without an API change, that is a solid option. Caveat:
- If cluster version were stored as a single string, we'd functionally have two version fields in the cluster CR (image name and introspected serialized version). Keeping them in sync across the cluster and its deployments would likely be more complex than this patch.
- To solve this we might want to consider storing it as a map[image]version. That's cumbersome, but the cluster CR is not authoritative on its deployments' running versions during an upgrade.

That being said, I still do think it's better to push this before the release branch. The crash-collector deployment should be properly labelled, and I'd vote to take on a bit of tech debt to avoid that UX snag.

**Comment by travisn** [comment]
Rather than something to completely avoid, I'd say a package variable is just the last resort for accessing something in the process. If you protect it with a mutex from multiple goroutines it will be safe. The preferred design pattern is just to pass parameters or encapsulate them in some other object so that objects don't accidentally trample on each other. In this case we have a singleton controller for each CR. A package variable is also a singleton, so we're in the same ballpark. An improvement could be an object that the controllers share, which contains the version info. The overhead of passing around this object that may not be worth it if it's a singleton anyway. It depends on how much of a purist you are. :)

For the implementation of storing it in the cluster CR, I was thinking the status would have something like this... When the CR is first created there is no status section. In this state, consumers that rely on the version info would have to wait until it's populated.
```
spec:
  cephVersion:
    image: ceph/ceph:v14.2.4
```

After the version is discovered, the operator saves it to the status. Consumers watching the CR could then read it.
```
spec:
  cephVersion:
    image: ceph/ceph:v14.2.4
status:
  cephVersion: 
    image: ceph/ceph:v14.2.4
    version: v14.2.4
```

In the future when it's time for an upgrade, the spec is updated, but the version info would be stale. Again, consumers of the version would have to wait until the status is updated to match the spec.
```
spec:
  cephVersion:
    image: ceph/ceph:v14.2.5
status:
  cephVersion: 
    image: ceph/ceph:v14.2.4
    version: v14.2.4
```

Finally, the operator would detect the upgraded status and everything would be in harmony again:
```
spec:
  cephVersion:
    image: ceph/ceph:v14.2.5
status:
  cephVersion: 
    image: ceph/ceph:v14.2.5
    version: v14.2.5
```

**Comment by egafford** [comment]
Sure; I definitely don't see wrapping the version map in a singleton as a step up. That'd just add complexity(, Java). Still, this is fundamentally different than a singleton that represents a functional component of the entire application or its configuration. It's data that's specific to a cluster or set of clusters, and that's the bit that I thought to be non-idiomatic here. We can make the argument that the clusterMap field of the ClusterController singleton is similar, but at least kvps there are keyed by namespace. It would take some quite awful practices for any given (image, version) kvp not to be a global constant and slippery slope arguments aren't always valid. Nonetheless I see where Blaine's coming from with his concerns about storing cluster data in global process memory as a precedent.

As to the status vs. spec synchronization, yes, as long as the status is always internally consistent then crash-collector and similarly managed deployments can detect its inconsistency with the spec. That can trigger the crash-collector and similarly managed deployments to wait to update to the new image until detect version completes. There's a bit of complexity there in that at present cluster.Status is only read or updated in a rather different flow (and goroutine) than the createInstance flow, and this would muddy that encapsulation a bit. Still, that's a solveable problem.

**Comment by travisn** [comment]
Ok I think I finally got it through my thick skull that the original argument was that we didn't want cluster-specific state stored as a package variable. If there are multiple clusters the singleton could cause issues. Agreed that it's not a pattern that we want to proliferate. In this case the mapping of the image to a version is global state that could be expected to be constant across clusters so this seems like a valid exception to that concern. 

A related optimization that could also be made is to skip running the detect version job if its version has already been detected. Orchestrations can be triggered for a number of reason and there's no reason to trigger a new detect version job if the image didn't change.
